DistroLauncher/ExitStatus.cpp

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "stdafx.h"
+#include <errno.h>
+#include <filesystem>
+
+namespace Oobe
+{
+
+    namespace
+    {
+        static const char* launcherCommandFilePath{"/run/launcher-command"};
+        VoidResult act(const KeyValuePairs& cmds);
+        VoidResult config(const KeyValuePairs& cmds);
+    }
+
+    void ExitStatusHandling()
+    {
+        std::string prefixedFilePath{"\\\\wsl$\\"};
+        auto distroName = Win32Utils::wide_string_to_utf8(DistributionInfo::Name);
+        // That should never fail, but if that happens it is a real bug.
+        if (!distroName.has_value()) {
+            wprintf(distroName.error().c_str());
+            return;
+        }
+        prefixedFilePath.append(distroName.value());
+        prefixedFilePath.append(launcherCommandFilePath);
+        std::ifstream launcherCmdFile;
+        launcherCmdFile.open(prefixedFilePath);
+        if (launcherCmdFile.fail()) {
+            if (errno != ENOENT) {
+                Helpers::PrintErrorMessage(HRESULT_FROM_WIN32(GetLastError()));
+            }
+            // File not found will be interpreted as there is nothing to do.
+            return;
+        }
+
+        auto parserResult = internal::parseExitStatusFile(launcherCmdFile);
+        if (parserResult.has_value()) {
+            // Finally take the actions and delete the file.
+            const KeyValuePairs& launcherCmds{parserResult.value()};
+            auto actionResult = act(launcherCmds);
+            auto configResult = config(launcherCmds);
+        } else {
+            wprintf(parserResult.error());
+        }
+
+        launcherCmdFile.close();
+        // We don't want that file existing after actions were taken.
+        std::filesystem::remove(prefixedFilePath);
+
+    }
+
+    namespace
+    {
+        bool ensureDistroStopped(unsigned int maxNoOfRetries);
```

Comment:
a separate declaration is not necessary because the body is below and nothing is referencing this in between?

Reply:
Indeed. In between moving functions around I ended up moving the calling one below the definition, thus this declaration is unnecessary. Thanks for pointing that out.